chalk-ir/src/lib.rs

```diff
@@ -1423,21 +1430,121 @@ impl<I: Interner> ProgramClauses<I> {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord, HasInterner)]
+pub struct ParameterKinds<I: Interner> {
+    interned: I::InternedParameterKinds,
+}
+
+impl<I: Interner> ParameterKinds<I> {
+    pub fn new(interner: &I) -> Self {
+        Self::from(interner, None::<ParameterKind<()>>)
+    }
+
+    pub fn interned(&self) -> &I::InternedParameterKinds {
+        &self.interned
+    }
+
+    pub fn from(
+        interner: &I,
+        parameter_kinds: impl IntoIterator<Item = ParameterKind<()>>,
+    ) -> Self {
+        ParameterKinds {
+            interned: I::intern_parameter_kinds(interner, parameter_kinds.into_iter()),
+        }
+    }
+
+    pub fn from_fallible<E>(
+        interner: &I,
+        parameter_kinds: impl IntoIterator<Item = Result<ParameterKind<()>, E>>,
+    ) -> Result<Self, E> {
+        let parameter_kinds = parameter_kinds
+            .into_iter()
+            .collect::<Result<Vec<ParameterKind<()>>, _>>()?;
+        Ok(Self::from(interner, parameter_kinds))
+    }
+
+    pub fn iter(&self, interner: &I) -> std::slice::Iter<'_, ParameterKind<()>> {
+        self.as_slice(interner).iter()
+    }
+
+    pub fn is_empty(&self, interner: &I) -> bool {
+        self.as_slice(interner).is_empty()
+    }
+
+    pub fn len(&self, interner: &I) -> usize {
+        self.as_slice(interner).len()
+    }
+
+    pub fn as_slice(&self, interner: &I) -> &[ParameterKind<()>] {
+        interner.parameter_kinds_data(&self.interned)
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord, HasInterner)]
+pub struct ParameterKindsWithUniverseIndex<I: Interner> {
```

Comment:
So in rustc we call this [`CanonicalVarInfos`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/infer/canonical/type.CanonicalVarInfos.html), and there is a struct [`CanonicalVarInfo`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/infer/canonical/struct.CanonicalVarInfo.html) that is roughly equivalent to `ParameterKind<UniverseIndex>`, although it includes a bit of extra information because we also introduce canonicals for placeholders. I think we want to unify these two approaches, which probably suggests a less awkward name for this struct -- something like `CanonicalVars`.

But I guess we should leave that to a follow-up PR.

Reply:
Done with `CanonicalVarKinds`.